lib/ex_aws/s3.ex

```diff
@@ -753,45 +814,65 @@ defmodule ExAws.S3 do
       <Days>#{number_of_days}</Days>
     </RestoreRequest>
     """
+
     request(:post, bucket, object, resource: "restore", params: params, body: body)
   end
 
   @type put_object_opts :: [
-    {:cache_control, binary}
-    | {:content_disposition, binary}
-    | {:content_encoding, binary}
-    | {:content_length, binary}
-    | {:content_type, binary}
-    | {:expect, binary}
-    | {:expires, binary}
-    | {:storage_class, :standard | :reduced_redundancy}
-    | {:website_redirect_location, binary}
-    | {:encryption, encryption_opts}
-    | {:meta, amz_meta_opts}
-    | acl_opts
-  ]
+          {:cache_control, binary}
+          | {:content_disposition, binary}
+          | {:content_encoding, binary}
+          | {:content_length, binary}
+          | {:content_type, binary}
+          | {:expect, binary}
+          | {:expires, binary}
+          | {:storage_class, :standard | :reduced_redundancy}
+          | {:website_redirect_location, binary}
+          | {:encryption, encryption_opts}
+          | {:meta, amz_meta_opts}
+          | acl_opts
+        ]
   @doc "Create an object within a bucket"
-  @spec put_object(bucket :: binary, object :: binary, body :: binary) :: ExAws.Operation.S3.t
-  @spec put_object(bucket :: binary, object :: binary, body :: binary, opts :: put_object_opts) :: ExAws.Operation.S3.t
+  @spec put_object(bucket :: binary, object :: binary, body :: binary) :: ExAws.Operation.S3.t()
+  @spec put_object(bucket :: binary, object :: binary, body :: binary, opts :: put_object_opts) ::
+          ExAws.Operation.S3.t()
   def put_object(bucket, object, body, opts \\ []) do
     request(:put, bucket, object, body: body, headers: put_object_headers(opts))
   end
 
   @doc "Create or update an object's access control policy"
-  @spec put_object_acl(bucket :: binary, object :: binary, acl :: [acl_opts]) :: ExAws.Operation.S3.t
+  @spec put_object_acl(bucket :: binary, object :: binary, acl :: [acl_opts]) ::
+          ExAws.Operation.S3.t()
   def put_object_acl(bucket, object, acl) do
-    headers = acl |> Map.new |> format_acl_headers
+    headers = acl |> Map.new() |> format_acl_headers
     request(:put, bucket, object, headers: headers, resource: "acl")
   end
 
-  @doc "Add a set of tags to an existing object"
+  @doc """
+  Add a set of tags to an existing object
+
+  ## Options
+
+  - `:version_id` - The versionId of the object that the tag-set will be added to.
+
+  """
   @spec put_object_tagging(
           bucket :: binary,
           object :: binary,
           tags :: Access.t(),
-          opts :: Keyword.t()
+          opts :: [option]
         ) :: ExAws.Operation.S3.t()
+        when option: {:version_id, binary} | {atom, binary}
   def put_object_tagging(bucket, object, tags, opts \\ []) do
+    {version_id, opts} = Keyword.pop(opts, :version_id)
```

Comment:
Take away `version_id` from options. It's okay because only `x-amz-expected-bucket-owner` and `content-md5` are only valid request headers.